File: leetcode/search_rotated_sorted_array_2/SearchRotatedSortedArray2.cpp
class Solution {
public:
    bool searchHelper(int A[], int l, int r, int target) {
        if (l > r) {
            return false; 
        }
        else {
            int mid = (l + r) / 2; 
            if (A[mid] == target) {
                return true; 
            }
            else if (A[mid] < target) {
                return searchHelper(A, mid + 1, r, target); 
            }
            else {
                return searchHelper(A, l, mid - 1, target); 
            }
        }
    }
    
    bool search(int A[], int n, int target) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        if (n == 0) {
            return false; 
        }
        
        int i = 1; 
        
        while (i < n && A[i] >= A[i - 1]) {
            i++; 
        }
        
        return searchHelper(A, 0, i - 1, target) || searchHelper(A, i, n - 1, target);  
    }
};

File: leetcode4/SearchRotatedSortedArray2.java
public class Solution {
    public boolean search(int[] A, int target) {
        int len = A.length; 
        
        if (len == 0) {
            return false; 
        }
        
        int index = -1; 
        
        for (int i = 0; i < len - 1; i++) {
            if (A[i] > A[i + 1]) {
                index = i; 
                break; 
            }
        }
        
        if (index == -1) {
            return binarySearch(A, 0, len - 1, target); 
        } else {
            if (A[0] <= target && A[index] >= target) {
                return binarySearch(A, 0, index, target); 
            } else {
                return binarySearch(A, index + 1, len -1, target); 
            }
        }
    }
    
    private boolean binarySearch(int[] A, int l, int r, int target) {
        if (l > r) {
            return false; 
        }
        
        int mid = (l + r) / 2; 
        if (A[mid] == target) {
            return true; 
        } else if (A[mid] < target) {
            return binarySearch(A, mid + 1, r, target); 
        } else {
            return binarySearch(A, l, mid - 1, target); 
        }
    }
}
